src/realm_dart_sync.cpp

```diff
@@ -33,19 +33,19 @@ RLM_API void realm_dart_http_request_callback(realm_userdata_t userdata, const r
     realm_http_request_t request_copy = request; // copy struct
```

Comment:
You don't need to copy the request here - you can `std::move` the request and create the copy inside the scheduler callback.

Reply:
I am not sure about that. The request is `const` arg so we should not `std::move` it

Reply:
`std::move`-ing doesn't guarantee that a move will occur, it only tells the compiler it _may_ move it.

Reply:
yes and accessing the old reference is undefined behavior. Hence I am hesitant to do what you suggest

Reply:
We're not accessing the old reference though, so I'm not sure I see the problem. If you do prefer to eagerly copy here, then we don't need to copy it again on line 45.

Reply:
My point is a `const` arg should not be modified. And by moving it we will invalidate it.

Reply:
on line 45 we do need to copy it again since it's not modifiable since its an rvalue.

Reply:
pardon me, on line 45 its not a copy. It's a move so as efficient as it can be.

Reply:
You can just make the argument not be `const`. Top-level `const` on arguments is ignored when deciding the function's type, so `void(foo)` and `void(const foo)` have the same type and can both be assigned to a `void(*)(foo)` function pointer. The `const` is only meaningful inside the implementation of the function as it makes the parameter value behave like a const object. In this case, that is probably something you don't want.

> std::move-ing doesn't guarantee that a move will occur, it only tells the compiler it may move it.

Sort of. `std::move` makes its result a (possibly const) rvalue. That then can affect overload resolution specifically to choose overloads that then move from the object. All of this is specifically defined, and no part of it is up to the compiler (except for the poorly named copy-elision optimization where it is allowed to completely skip both copies and moves in some cases).

> yes and accessing the old reference is undefined behavior.

Not quite, although it is perhaps a good rule of thumb to pretend this is the case most of the time. A) it isn't the `std::move` that causes this affect, it is whether a move actually occurs. B) For all standard library types, you are guaranteed that a moved-from object will be left in a "valid but unspecified state". This means that you can call any methods that don't have preconditions. For user-defined types, unless otherwise stated, it is best to assume that you can only destroy the object or assign over it. "Named assignments" such as `clear()` or `reset()` are also generally considered acceptable.

(As I said, only using `std::move` on object you don't touch again is a good safe rule of thumb if you don't want to think too hard. But this thread was already getting into the weeds, so I wanted to make sure there was a precise statement of what the real rules are)

Reply:
Thanks @RedBeard0531. Good insights. I was with the impression that `const args` will help the compiler in someway optimize the function invocation. I will remove the const from the arg there.